src/matchmaker.rs

```diff
@@ -0,0 +1,358 @@
+use std::collections::HashMap;
+
+#[derive(Eq, PartialEq)]
+enum RangeOperator {
+    GT,
+    LT,
+    GEQ,
+    LEQ,
+}
+
+enum QueryType {
+    Term(String),
+    Range { value: i32, operator: RangeOperator },
+}
+
+#[derive(Eq, PartialEq)]
+enum Boolean {
+    Required,
+    Optional,
+    Excluded,
+}
+
+pub struct Matchmaker {
+    pub min_count: u32,
+    pub max_count: u32,
+    string_properties: HashMap<String, String>,
+    numeric_properties: HashMap<String, f64>,
+    pub query: String,
+}
+
+pub struct QueryItemBuilder<'a> {
+    matchmaker: &'a mut Matchmaker,
+    property: &'a str,
+    query_type: Option<QueryType>,
+    boolean: Boolean,
+    boost: i64,
+}
+
+impl<'a> QueryItemBuilder<'a> {
+    pub fn term(&mut self, term: &str) -> &mut Self {
+        self.query_type = Some(QueryType::Term(term.to_owned()));
+        self
+    }
+
+    pub fn gt(&mut self, value: i32) -> &mut Self {
+        self.query_type = Some(QueryType::Range {
+            operator: RangeOperator::GT,
+            value,
+        });
+        self
+    }
+
+    pub fn geq(&mut self, value: i32) -> &mut Self {
+        self.query_type = Some(QueryType::Range {
+            operator: RangeOperator::GEQ,
+            value,
+        });
+        self
+    }
+
+    pub fn lt(&mut self, value: i32) -> &mut Self {
+        self.query_type = Some(QueryType::Range {
+            operator: RangeOperator::LT,
+            value,
+        });
+        self
+    }
+
+    pub fn leq(&mut self, value: i32) -> &mut Self {
+        self.query_type = Some(QueryType::Range {
+            operator: RangeOperator::LEQ,
+            value,
+        });
+        self
+    }
+
+    pub fn required(&mut self) -> &mut Self {
+        self.boolean = Boolean::Required;
+        self
+    }
+
+    pub fn excluded(&mut self) -> &mut Self {
+        self.boolean = Boolean::Excluded;
+        self
+    }
+
+    pub fn boost(&mut self, amount: i64) -> &mut Self {
+        self.boost = amount;
+        self
+    }
+
+    pub fn build(&'a mut self) -> &'a mut Matchmaker {
```

Comment:
With `build` like this we do not have type-guarantee that Matchmaker is not used without a `build`.
And also it is not really obvious from types that `.build()` should be called at all - `Matchmaker::new().boost(5)` already gave `Matchmaker`, why bother with `build()`?

So I think the possible solutions there:
- make sure that `.build()` is not needed. Matchmaker should be good to go after any `.whatever()` call
- separate `MatchmakerBuilder` and `Matchmaker`. The only way to get valid `Matchmaker` would be - through `MatchmakerBuilder`.

For reference: https://doc.rust-lang.org/1.0.0/style/ownership/builders.html